clarity_ext/repository/step_repository.py

```diff
@@ -169,3 +173,168 @@ def get_process_type(self):
+    for input, output, output_generation_type in wrappable_pairs:
+        input  # genologics Artifact, either stateful or stateless
+        output  # genologics Artifact
+
+    The validation compares the stateless and statefull xml representation,
+    and discards any discrepancies in the qc-flag
+    """
+
+    MODE_STATELESS = "stateless"
+    MODE_STATEFUL = "stateful"
+
+    def __init__(self, session, input_output_maps, state_mode):
+        # Note, the force flag should be set to False as soon as the test period is over!
+        # Develop-1177
+        self.force = True
+        self.session = session
+        self.input_output_maps = input_output_maps
+        self.state_mode = state_mode
+        self.pairs = list()
+        stateful_artifacts = self._fetch_stateful()
+        stateless_artifacts = self._fetch_stateless()
+
+        self.pairs = self._assemble_pairs(stateful_artifacts, stateless_artifacts)
+
+    def validate(self):
+        errors = list()
+        for pair in self.pairs:
+            errors.extend(pair.validate(self.session.api))
+        return errors
+
+    def __iter__(self):
+        self.counter = 0
+        return self
+
+    def next(self):
+        return self.__next__()
+
+    def __next__(self):
+        if self.counter == len(self.pairs):
+            raise StopIteration
+        pair = self.pairs[self.counter]
+        self.counter += 1
+        return iter((pair.input_artifact, pair.output_artifact, pair.output_generation_type))
+
+    def _assemble_pairs(self, stateful_artifacts, stateless_artifacts):
+        pairs = list()
+        stateful_by_uri = {artifact.uri: artifact for artifact in stateful_artifacts}
+        stateless_by_lims_id = {artifact.id: artifact for artifact in stateless_artifacts}
+        for input, output in self.input_output_maps:
+            stateful_input = stateful_by_uri[input["uri"].uri]
+            fetched_input = ArtifactRepresentation(
+                stateful_representation=stateful_input,
+                stateless_representation=stateless_by_lims_id[stateful_input.id]
+            )
+            stateful_output = stateful_by_uri[output["uri"].uri]
+            fetched_output = ArtifactRepresentation(
+                stateful_representation=stateful_output,
+                stateless_representation=stateless_by_lims_id[stateful_output.id]
+            )
+            pair = Pair(
+                input=fetched_input,
+                output=fetched_output,
+                output_generation_type=output["output-generation-type"],
+                state_mode=self.state_mode
+            )
+            pairs.append(pair)
+
+        return pairs
+
+    def _fetch_stateful(self):
+        artifact_keys = set()
+        for input, output in self.input_output_maps:
+            artifact_keys.add(input["uri"])
+            artifact_keys.add(output["uri"])
+
+        return self.session.api.get_batch(artifact_keys, force=self.force)
```

Comment:
This is a minor point, as it just concerns two calls. But can we fetch all in one batch call? I.e. merge `_fetch_stateful` and `_fetch_stateless`? Do you think that's unnecessary micro-optimization at this point?

Reply:
I suspect there is some trickery with the cache and the state-flag. I would not be surprised if the state part is remove before adding it to the cache. (I had to add the force flag in order to fetch the artifacts a second time...)

Reply:
Yes, I was also wondering if this would be unnecessary complication :+1: